crates/storage/provider/src/providers/mod.rs

```diff
@@ -334,7 +334,14 @@ where
 
     fn history_by_block_number(&self, block_number: BlockNumber) -> Result<StateProviderBox<'_>> {
         trace!(target: "providers::blockchain", ?block_number, "Getting history by block number");
-        self.database.history_by_block_number(block_number)
+
+        let latest = self.database.best_block_number()?;
+        // Return an error if requested block is not synced yet
```

Comment:
this still does another read :)

I know all of these traits are a bit confusing, but take a look at:

https://github.com/paradigmxyz/reth/blob/f00dc091deef3c2abd5c336eab0ffa6681596336/crates/storage/provider/src/providers/mod.rs#L138-L140

this should make it obvious what the difference is.

I also opened #2830 which is somewhat related to this entire thing.

please also add a few docs here (see issue) why we're doing this check